### PR TITLE
CDAP-3521 Recording lineage information for a dataset accessed by a flowlet

### PR DESCRIPTION
**Pull request by poornachandra**
Only recording lineage for datasets from flowlets in this PR to keep changes minimal. I'll add recording for streams and other program types in an upcoming PR.

Also, the following TODOs will be fixed in an upcoming PR:
1. Avoid duplicate writes of lineage for a dataset instance.
2. Add complete metadata and tags when recording lineage.

The PR looks big due to refactoring of tests and new classes for serializing lineage over HTTP.

### CODE REVIEW

**Comment by poornachandra** [comment]
Build - http://builds.cask.co/browse/CDAP-DUT2810-8

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ */
+public class BasicLineageWriter implements LineageWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(BasicLineageWriter.class);
+  private static final Gson GSON = new Gson();
+
+  private final BusinessMetadataStore businessMetadataStore;
+  private final LineageStore lineageStore;
+
+  @Inject
+  public BasicLineageWriter(BusinessMetadataStore businessMetadataStore, LineageStore lineageStore) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java]: package private

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                        Id.NamespacedId component) {
+    // TODO: avoid duplicate writes for a dataset instance
+    String metadata = gatherMetadata(run, datasetInstance);
+    LOG.debug("Writing access for run {}, dataset {}, accessType {}, component {}, metadata = {}",
+              run, datasetInstance, accessType, component, metadata);
+    lineageStore.addAccess(run, datasetInstance, accessType, metadata, component);
+  }
+
+  private String gatherMetadata(Id.Run run, Id.DatasetInstance datasetInstance) {
+    // TODO: add app metadata and tags
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java]: this is available, right? just that you need to create an object to represent all this info? Makes sense to do in another PR

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+ */
+public class LineageWriterDatasetFramework implements DatasetFramework, ProgramContextAware {
+  private final DatasetFramework delegate;
+  private final LineageWriter lineageWriter;
+
+  private final AtomicReference<Id.Run> runRef = new AtomicReference<>();
+  private final AtomicReference<Id.NamespacedId> componentIdRef = new AtomicReference<>();
+
+  @Inject
+  public LineageWriterDatasetFramework(
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: package private

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+package co.cask.cdap.metadata.serialize;
+
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Class to serialize {@link co.cask.cdap.data2.metadata.lineage.Relation}.
+ */
+public class RelationProto {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java]: should these classes move to ``cdap-proto``?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java [hunk 5]
```diff
@@ -23,16 +23,22 @@
...
 import com.google.inject.Scopes;
 import com.google.inject.assistedinject.FactoryModuleBuilder;
+import com.google.inject.name.Names;
 
 /**
  * DataSets framework bindings
  */
 public class DataSetsModules extends RuntimeModule {
 
+  public static final String BASIC_DATASET_FRAMEWORK = "basicDatasetFramework";
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java]: should we call this ``delegateDatasetFramework`` to indicate that the ``LineageDatasetFramework`` will delegate to these classes?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/Lineage.java [hunk 6]
```diff
@@ -27,27 +26,15 @@
  */
 public class Lineage {
   private final Set<Relation> relations;
-  private final Set<Id.Program> programs;
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/Lineage.java]: how come these two were removed? Just the fact that they can be derived from the set of relations? I think they provided a nice utility, instead of users having to compute this themselves?

**Comment by bdmogal** [comment]
@poornachandra the approach makes sense to me and is easy to understand. Only major comments  about moving the proto classes to ``cdap-proto``, and the removal of programs and data from the ``Lineage`` class.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java (hunk 4 above)
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java]: Since these classes are only used for Lineage now, let's leave it here. It should be easy to move them when required, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java (hunk 5 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java]: `basicDatasetFramework` is used by classes other than `LineageDatasetFramework`, for them delegate does not add any meaning I think.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/Lineage.java (hunk 6 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/Lineage.java]: Right now they are not being used anywhere. Having them only increased the size of code, hence removed them. It should be easy to add them back if required later.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java (hunk 2 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java]: Yes.

**Comment by poornachandra** [comment]
@bdmogal Thanks for the review. Addressed your comments. Can you please take another look?

Also, can I make the package-private changes in the upcoming PR?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java (hunk 4 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java]: ``LineageProto`` is the output of the lineage REST API, though, right? Its ok to do it in another PR, but we should do it in 3.2?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java (hunk 5 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java]: Ok, maybe I missed it. Will the ``DatasetFramework`` named as ``basicDatasetFramework`` be used by anything other than the ``LineageDatasetFramework``?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/Lineage.java (hunk 6 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/Lineage.java]: Sounds good.

**Comment by bdmogal** [comment]
Ok to make changes in another PR since multiple PRs depend on each other here. LGTM.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java (hunk 5 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java]: Yes, it will be used by classes that `LineageDatasetFramework` depends on, such as `BusinessMetadataStore` and `LineageStore` to prevent circular dependency.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java (hunk 4 above)
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java]: If all classes used as output of REST API have to go to cdap-proto then I'll move them.

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java]: yes, that's the convention. It's ok to create a move-only PR after this one though.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java (hunk 5 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java]: Ok clearly I missed that. **basic** is reasonable.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java (hunk 4 above)
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java]: Will do.

**Comment by poornachandra** [comment]
Build passed, merging.
